src/task.c

```diff
@@ -640,72 +661,69 @@ task_enqueue(task_system *ts,
       }
    }
    platform_condvar_signal(&group->cv);
-   return platform_condvar_unlock(&group->cv);
+   return task_group_unlock(group);
 }
 
+/*
+ * Run a task if the number of waiting tasks is at least min_backlog.
+ */
 static platform_status
-task_group_perform_one(task_group *group, uint64 min_backlog)
+task_group_perform_one(task_group *group, uint64 queue_scale_percent)
 {
    platform_status rc;
    task           *assigned_task = NULL;
-   if (group->current_waiting_tasks <= min_backlog) {
+
+   /* We do the queue size comparison in this round-about way to avoid
+      integer overflow. */
+   if (queue_scale_percent
+       && 100 * group->current_waiting_tasks / queue_scale_percent
+             < group->bg.num_threads)
+   {
       return STATUS_TIMEDOUT;
    }
 
-   platform_thread_cleanup_push((void (*)(void *))platform_condvar_unlock,
-                                &group->cv);
-   rc = platform_condvar_lock(&group->cv);
+   rc = task_group_lock(group);
    if (!SUCCESS(rc)) {
-      goto out;
+      return rc;
    }
 
-   assigned_task = task_group_get_one_task(group);
+   assigned_task = task_group_get_next_task(group);
 
-out:
+   /* It is important to update the current_executing_tasks while
+      holding the lock. The reason is that, if we release the lock
+      before updating current_executing_tasks, then another thread
+      might observe that both the number of enqueued tasks and the
+      number of executing tasks are both 0, and hence that the system
+      is quiescent, even though it is not. */
    if (assigned_task) {
       __sync_fetch_and_add(&group->current_executing_tasks, 1);
    }
 
-   platform_condvar_unlock(&group->cv);
-   platform_thread_cleanup_pop(0);
+   task_group_unlock(group);
 
    if (assigned_task) {
       const threadid tid = platform_get_tid();
       group->stats[tid].total_fg_task_executions++;
       task_group_run_task(group, assigned_task);
       __sync_fetch_and_sub(&group->current_executing_tasks, 1);
+      platform_free(group->ts->heap_id, assigned_task);
+   } else {
+      rc = STATUS_TIMEDOUT;
    }
 
    return rc;
 }
 
 /*
- * task_perform_one() - Checks if there is a task to run and runs it.
- * Returns ETIMEDOUT if it didn't run any task;
+ * Perform a task only if there are more waiting tasks than bg
+ * threads.
```

Comment:
Request to clarify this comment: "_if there are more waiting tasks than bg threads._" ... is accurate if the queue_scale_percent is at its default, 100%.

But the actual behaviour implemented by `task_group_perform_one()` to decide "if_needed" semantic depends on this new param added here.

Sorry to be nit picky ... but it will help correct the comment and somehow cast the behaviour of "if_needed" in terms of this new argument, so that the lay reader is not mislead.